bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs

```diff
@@ -0,0 +1,322 @@
+	/// 	above will cause the conversion to fail
+	/// e. Currently, `AliasOrigin` is always required, can distinguish the V2 process from V1.
+	/// 	it's required also for dispatching transact from that specific origin.
+	/// f. SetTopic is required for tracing the message all the way along.
+	pub fn convert(&mut self) -> Result<Message, XcmConverterError> {
+		// Get fee amount
+		let fee_amount = self.extract_remote_fee()?;
+
+		// Get ENA reserve asset from WithdrawAsset.
+		let mut enas =
+			match_expression!(self.peek(), Ok(WithdrawAsset(reserve_assets)), reserve_assets);
+		if enas.is_some() {
+			let _ = self.next();
+		}
+
+		// Get PNA reserve asset from ReserveAssetDeposited
+		let pnas = match_expression!(
+			self.peek(),
+			Ok(ReserveAssetDeposited(reserve_assets)),
+			reserve_assets
+		);
+		if pnas.is_some() {
+			let _ = self.next();
+		}
+
+		// Try to get ENA again if it is after PNA
+		if enas.is_none() {
+			enas =
+				match_expression!(self.peek(), Ok(WithdrawAsset(reserve_assets)), reserve_assets);
+			if enas.is_some() {
+				let _ = self.next();
+			}
+		}
+		// Check AliasOrigin.
+		let origin_location = match_expression!(self.next()?, AliasOrigin(origin), origin)
+			.ok_or(AliasOriginExpected)?;
+		let origin = AgentIdOf::convert_location(origin_location).ok_or(InvalidOrigin)?;
+
+		let (deposit_assets, beneficiary) = match_expression!(
+			self.next()?,
+			DepositAsset { assets, beneficiary },
+			(assets, beneficiary)
+		)
+		.ok_or(DepositAssetExpected)?;
+
+		// assert that the beneficiary is AccountKey20.
+		let recipient = match_expression!(
+			beneficiary.unpack(),
+			(0, [AccountKey20 { network, key }])
+				if self.network_matches(network),
+			H160(*key)
+		)
+		.ok_or(BeneficiaryResolutionFailed)?;
+
+		// Make sure there are reserved assets.
+		if enas.is_none() && pnas.is_none() {
+			return Err(NoReserveAssets);
+		}
+
+		let mut commands: Vec<Command> = Vec::new();
+
+		// ENA transfer commands
+		if let Some(enas) = enas {
+			commands.append(&mut self.extract_ethereum_native_assets(
+				enas,
+				deposit_assets,
+				recipient,
+			)?);
+		}
+
+		// PNA transfer commands
+		if let Some(pnas) = pnas {
+			commands.append(&mut self.extract_polkadot_native_assets(
+				pnas,
+				deposit_assets,
+				recipient,
+			)?);
+		}
+
+		// Transact commands
+		let transact_call = match_expression!(self.peek(), Ok(Transact { call, .. }), call);
+		if let Some(transact_call) = transact_call {
+			let _ = self.next();
+			let transact =
+				ContractCall::decode_all(&mut transact_call.clone().into_encoded().as_slice())
+					.map_err(|_| TransactDecodeFailed)?;
+			match transact {
+				ContractCall::V1 { target, calldata, gas, value } => commands
+					.push(Command::CallContract { target: target.into(), calldata, gas, value }),
+			}
+		}
+
+		// ensure SetTopic exists
+		let topic_id = match_expression!(self.next()?, SetTopic(id), id).ok_or(SetTopicExpected)?;
```

Comment:
@yrong @alistair-singh 

I think explicitly requiring a topic is counter-productive. If the user is forced to provide a topic they may well just provide zero bytes as the topic. If the topic isn't present maybe we should generate one on the fly. This matches the behavior in the inbound direction.

Can use [unique](https://paritytech.github.io/polkadot-sdk/master/frame_system/fn.unique.html) for this, providing the input xcm as the argument.

Reply:
SovereignPaidRemoteExporter automatically inserts the topic if it is not present in the xcm.

https://github.com/paritytech/polkadot-sdk/blob/9dbcc0e4620380c11d2efb91bfb942927dcd6eb5/polkadot/xcm/xcm-builder/src/universal_exports.rs#L368-L373

So this piece of code always expects asset hub to insert it, so I think it is ok to fail if it's not here. I mean we could make this code more robust by generating a topic here if it does not exist, but practically would only save us in the event that we use some other exporter that does not do this.

Reply:
https://linear.app/snowfork/issue/SNO-1250#comment-b18d0916

I checked it before, seems all parachains have added the `WithUniqueTopic`(sender end) and `TrailingSetTopicAsId`(receiver end) wrappers, which is the recommended configuration by Parity.

So I'm not too worried about that.

Reply:
https://github.com/acatangiu/polkadot-sdk/pull/5